clusterloader2/pkg/measurement/util/phase_latency.go

```diff
@@ -0,0 +1,123 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"k8s.io/klog"
+)
+
+// Transition describe transition between two phases.
+type Transition struct {
+	From      string
+	To        string
+	Threshold time.Duration
+}
+
+// ObjectTransitionTimes stores beginning time of each phase.
+// It can calculate transition latency between phases.
+type ObjectTransitionTimes struct {
+	name  string
+	times map[string]map[string]time.Time
+}
+
+//NewObjectTransitionTimes creates new ObjectTransitionTimes instance.
+func NewObjectTransitionTimes(name string) *ObjectTransitionTimes {
+	return &ObjectTransitionTimes{
+		name:  name,
+		times: make(map[string]map[string]time.Time),
+	}
+}
+
+// Set sets time of given phase for given key.
+func (o *ObjectTransitionTimes) Set(key, phase string, t time.Time) {
+	if _, exists := o.times[key]; !exists {
+		o.times[key] = make(map[string]time.Time)
+	}
+	o.times[key][phase] = t
+}
+
+// Get returns time of given phase for given key.
+func (o *ObjectTransitionTimes) Get(key, phase string) (time.Time, bool) {
```

Comment:
I'm not sure you really need Get() for anything.

The users of that type should only be setting, and the implementation here should handle correctly partially-filled objects.

Reply:
I consider it convenient - it allows (for example) to check if previous phase exists etc... + I'm using it for pod scheduling event filtering. Without this filtering there could be a lot of missing phase time logged.

Reply:
I'm not convinced, but it's minor enough that I can live with it.